File: ch04/slice_alloc.zig
const std = @import("std");
pub fn main() !void {
    var allocator = std.heap.page_allocator; // allocate using a page allocator
    // what's done in createSlice...create our pointer for 5 i32 elements
    var ptr = try allocator.alloc(i32, 5);
    var slice: []i32 = &[_]i32{}; // create a slice using the allocated memory
    slice = ptr;
    slice.len = 5; // set the length to 5
    // what's done in main...assign values to the elements of the slice
    for(slice) |_, i|{slice[i] = @intCast(i32, i);}
    // Print the values of the slice
    for (slice) |element| {
        std.debug.print("{} ", .{element});
    }
    // what's done in freeSlice...free the pointer we allocated
    allocator.free(ptr);
}

File: ch04/slice_alloc.c
#include<stdio.h>
#include<stdlib.h>
typedef struct { // A structure to represent a slice in Zig
    int* ptr; // a dynamic array
    size_t len; // the length the slice takes
} Slice;
// Creates a new slice given a specified length
Slice createSlice(size_t length){
    // allocate memory for our pointer given the length 
    int* ptr = malloc(length * sizeof(int));
    // create an unitialized slice 
    Slice slice; 
    // set the ptr of the slice to `ptr'
    slice.ptr = ptr; 
    // set the length to our given length
    slice.len = length; 
    return slice;
}
// Frees the slice and makes it empty 
void freeSlice(Slice* slice){
    free(slice->ptr); // free the pointer in our slice 
    slice->ptr = NULL; // set it back to NULL
    slice->len = 0; // set the length to 0 
}

int main(){
    // create a slice that allocates a size of 5
    Slice slice = createSlice(5); 
    // assign values to the elements of the slice 
    for(size_t i = 0; i < slice.len; i++){ slice.ptr[i] = i + 1; }
    // print the values 
    for(size_t i = 0; i < slice.len; i++){ printf("%d ", slice.ptr[i]); }
    printf("\n");
    // free the slice at the end of the scope 
    freeSlice(&slice);
    return 0;
}